fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/TransactionImpl.java

```diff
@@ -30,40 +37,69 @@
  */
 public class TransactionImpl implements Transaction {
 
+    private final static Logger log = LoggerFactory.getLogger(TransactionImpl.class);
+
     final String id;
 
+    final TransactionManagerImpl txManager;
+
+    final Duration DEFAULT_TIMEOUT = ofMinutes(3);
+
     boolean shortLived = true;
 
-    TransactionImpl(final String id) {
+    Instant expiration;
+
+    boolean expired = false;
+
+    boolean rolledback = false;
+
+    boolean commited = false;
+
+    protected TransactionImpl(final String id, final TransactionManagerImpl txManager) {
         if (id == null || id.isEmpty()) {
             throw new IllegalArgumentException("Transaction id should not be empty!");
         }
         this.id = id;
+        this.txManager = txManager;
+        this.expiration = Instant.now().plus(timeout());
     }
 
     @Override
     public void commit() {
-        // Prepare Persistence Transactions
-        // Commit Persistence Transactions
+        failIfExpired();
+        failIfRolledback();
+        if(this.commited) {
+            return;
+        }
+        try {
+            log.debug("Commiting transaction {}", id);
+            this.getPersistentSession().commit();
+            this.commited = true;
+        } catch (final PersistentStorageException ex) {
+            throw new RepositoryRuntimeException("failed to commit transaction " + id, ex);
```

Comment:
If a commit fails, should the transaction still be usable? From the work that Danny has been doing, it seems like we are going to want to trigger an automatic rollback on failure, which would indicate that the Transaction should no longer be usable for requests. This might be resolvable by simply calling `rollback()` in this catch block before throwing the error.

I have the feeling that we'll want more precise exceptions being thrown at some point so we can tell conflicts from server errors, but I don't think there's anything you can do about it at this point.